sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala

```diff
@@ -722,14 +724,246 @@ object JdbcUtils extends Logging {
   }
 
   /**
+   * Check whether a table exists in a given database
+   *
+   * @return True if the table exists.
+   */
+  @transient
+  def checkTableExists(targetDb: String, tableName: String): Boolean = {
+    val dbc: Connection = DriverManager.getConnection(targetDb)
+    val dbm = dbc.getMetaData()
+    // Check if the table exists. If it exists, perform an upsert.
+    // Otherwise, do a simple dataframe write to the DB
+    val tables = dbm.getTables(null, null, tableName, null)
+    val exists = tables.next() // Returns false if next does not exist
+    dbc.close()
+    exists
+  }
+
+  // Provide a reasonable starting batch size for database operations.
+  private val DEFAULT_BATCH_SIZE: Int = 200
+
+  // Limit the number of database connections. Some DBs suffer when there are many open
+  // connections.
+  private val DEFAULT_MAX_CONNECTIONS: Int = 50
```

Comment:
Well, since Spark 2.1, we already provide the parm for limiting the max num of concurrent JDBC connection when inserting data to JDBC tables. The parm is `numPartitions`.